### PR TITLE
Experiment with sorting

### PR DESCRIPTION
**Pull request by alexey-milovidov**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Avoid processing already sorted data.

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
Extreme case.

New:
```
milovidov-desktop :) SELECT number FROM numbers(100000000) ORDER BY number LIMIT 90000000, 10

┌───number─┐
│ 90000000 │
│ 90000001 │
│ 90000002 │
│ 90000003 │
│ 90000004 │
│ 90000005 │
│ 90000006 │
│ 90000007 │
│ 90000008 │
│ 90000009 │
└──────────┘

10 rows in set. Elapsed: 0.660 sec. Processed 100.01 million rows, 800.08 MB (151.45 million rows/s., 1.21 GB/s.)
```

Old:
```
milovidov-desktop :) SELECT number FROM numbers(100000000) ORDER BY number LIMIT 90000000, 10

ClickHouse server version is older than ClickHouse client. It may indicate that the server is out of date and can be upgraded.

┌───number─┐
│ 90000000 │
│ 90000001 │
│ 90000002 │
│ 90000003 │
│ 90000004 │
│ 90000005 │
│ 90000006 │
│ 90000007 │
│ 90000008 │
│ 90000009 │
└──────────┘

10 rows in set. Elapsed: 4.579 sec. Processed 100.03 million rows, 800.21 MB (21.84 million rows/s., 174.75 MB/s.)
```

File: src/Interpreters/sortBlock.cpp [hunk 1]
```diff
@@ -194,27 +189,16 @@ void getBlockSortPermutationImpl(const Block & block, const SortDescription & de
...
-    if (!block)
+        if (limit && block.rows() > limit)
+            for (auto & elem : block)
+                elem.column = elem.column->cut(0, limit);
         return;
+    }
 
     IColumn::Permutation permutation;
-    getBlockSortPermutationImpl(block, description, IColumn::PermutationSortStability::Stable, 0, permutation);
+    getBlockSortPermutationImpl(block, description, IColumn::PermutationSortStability::Unstable, limit, permutation);
```
Comment (devcrafter) [inline src/Interpreters/sortBlock.cpp:201]: `getBlockSortPermutationImpl()` can be made `static` since used only in this file

**Review by devcrafter — APPROVED** [review]
LGTM, just need to check performance tests

File: src/Interpreters/sortBlock.cpp (hunk 1 above)
Comment (alexey-milovidov) [reply src/Interpreters/sortBlock.cpp:179]: It was already in the anonymous namespace.

**Comment by alexey-milovidov** [comment]
It's worse in case of a small LIMIT due to a large number of virtual calls:
https://s3.amazonaws.com/clickhouse-test-reports/46558/f66a7702053cac9128085ff628ffb831196deacb/performance_comparison_[3/4]/report.html

**Comment by alexey-milovidov** [comment]
We need a batch version of `isAlreadySorted` in the `IColumn` interface.

Also, @kitaisreal proposed a good idea: we can do sorting to obtain the permutation, but then look if the permutation is identity, and avoid permitting in this case.

Also, `pdqsort` wins in the case if the data is already reverse sorted. We might need to speculate on this case as well.

**Comment by alexey-milovidov** [comment]
Merged in #50697.

**Pull request closed without merging**